### PR TITLE
Reorder smoother update responsibility

### PR DESCRIPTION
**Pull request by ManInFez**
**Issue**
Resolves #my_issue

**Approach**
_Short description of the approach_

## Pre review checklist

- [ ] Added appropriate labels

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

**Comment by ManInFez** [comment]
Test ert please!

**Comment by ManInFez** [comment]
Test ert please!

**Comment by ManInFez** [comment]
Test ert please!

**Comment by ManInFez** [comment]
test ert please!

**Comment by ManInFez** [comment]
Test ert please!

File: libres/lib/analysis/update.cpp [hunk 1]
```diff
@@ -67,6 +66,46 @@ typedef struct {
         *node_info; /* mutable: For the serialization of one node */
 } serialize_info_type;
 
+class update_data_type {
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:69]: Considering adding docstring.
Something along the lines of "Container for all matrices needed for an update step"

File: libres/lib/analysis/update.cpp [hunk 2]
```diff
@@ -67,6 +66,46 @@ typedef struct {
...
+            A_with_rowscaling_in,
+        const bool_vector_type *obs_mask_in)
+        : obs_mask(obs_mask_in) {
+        S = S_in;
+        E = E_in;
+        D = D_in;
+        R = R_in;
+        A = A_in;
+        A_with_rowscaling = A_with_rowscaling_in;
+        valid = true;
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:85]: What is valid?

Comment (joakim-hove) [reply libres/lib/analysis/update.cpp:85]: It is the entire `update_data_type` instance which is valid - or not. The `valid` value is based on whether we have any remaining observations - if the number of observations is zero (and hence no update should be performed), we have `valid = false`. A better name would maybe be `have_observations`

File: libres/lib/analysis/update.cpp [hunk 3]
```diff
@@ -354,7 +382,8 @@ void run_analysis_update_with_rowscaling(
 
     ert::utils::Benchmark benchmark(logger,
                                     "run_analysis_update_with_rowscaling");
-    assert(parameters.size() > 0);
+    if (parameters.size() == 0)
+        return;
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:386]: Should this throw an error or warning of sorts?

File: libres/lib/analysis/update.cpp [hunk 4]
```diff
@@ -446,96 +482,123 @@ void assert_size_equal(int ens_size, const bool_vector_type *ens_mask) {
...
-bool smoother_update(std::vector<int> step_list,
-                     const local_updatestep_type *updatestep,
+update_data_type make_update_data(enkf_fs_type *source_fs,
+                                  enkf_fs_type *target_fs, enkf_obs_type *obs,
+                                  ensemble_config_type *ensemble_config,
+                                  const analysis_config_type *analysis_config,
+                                  bool_vector_type *ens_mask,
+                                  local_ministep_type *ministep,
+                                  rng_type *shared_rng, FILE *log_stream) {
+    /*
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:503]: Is this docstring outdated?

File: libres/lib/analysis/update.cpp (hunk 3 above)
Comment (joakim-hove) [reply libres/lib/analysis/update.cpp:386]: I would say no - this is a totally legitimate situation; one could move to the check to calling scope and if `parameters.size() == 0` not call the function at all.

File: libres/lib/analysis/update.cpp [hunk 5]
```diff
@@ -446,96 +482,123 @@ void assert_size_equal(int ens_size, const bool_vector_type *ens_mask) {
...
+    assert_matrix_size(R, "R", active_obs_size, active_obs_size);
+    const bool_vector_type *obs_mask = obs_data_get_active_mask(obs_data);
+    obs_data_scale(obs_data, S, E, D, R, nullptr);
+
+    auto row_scaling_parameters = load_row_scaling_parameters(
+        target_fs, ensemble_config, iens_active_index, active_ens_size,
+        ministep);
+
+    local_ministep_add_obs_data(ministep,
+                                obs_data); // This is not correct conceptually
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:565]: What is not correct conceptually?

File: libres/lib/analysis/update.cpp (hunk 2 above)
Comment (dafeda) [reply libres/lib/analysis/update.cpp:85]: From what you write I would suggest `has_observations` and a short descriptive comment.

File: libres/lib/analysis/update.cpp (hunk 3 above)
Comment (dafeda) [reply libres/lib/analysis/update.cpp:386]: If we move the check to the calling scope, would it then not make sense to throw an error if `run_analysis_update_with_rowscaling` is called with parameter.size() == 0?

File: libres/lib/analysis/update.cpp (hunk 5 above)
Comment (ManInFez) [reply libres/lib/analysis/update.cpp:565]: ministep should only hold the configuration objects, like selected observations and parameters for the update step, not the actual data.

File: libres/lib/analysis/update.cpp (hunk 2 above)
Comment (ManInFez) [reply libres/lib/analysis/update.cpp:85]: valid whether is has the data for doing an update step

File: libres/lib/analysis/update.cpp (hunk 5 above)
Comment (dafeda) [reply libres/lib/analysis/update.cpp:565]: That's a better comment I think.

File: libres/lib/analysis/update.cpp (hunk 3 above)
Comment (joakim-hove) [reply libres/lib/analysis/update.cpp:386]: No - the early return here is just a convenience, it is perfectly legitimate to run through this function with `paramaters = []` it is just a minor waste.

File: libres/lib/analysis/update.cpp (hunk 5 above)
Comment (joakim-hove) [reply libres/lib/analysis/update.cpp:565]: To me it looks like this is (partly ?) addressed in #2923

File: libres/lib/analysis/update.cpp (hunk 3 above)
Comment (dafeda) [reply libres/lib/analysis/update.cpp]: Alright, I don't like it but can definitely live with it.

File: libres/lib/analysis/update.cpp (hunk 1 above)
Comment (ManInFez) [reply libres/lib/analysis/update.cpp:69]: I have added it but it does not show up in the diff :shrug:

**Review by dafeda — APPROVED** [review]
Good stuff 👍

**Pull request merged**